cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordCounterService.java

```diff
@@ -16,34 +16,90 @@
 
 package io.cdap.cdap.internal.app.services;
 
+import com.google.common.util.concurrent.AbstractScheduledService;
+import com.google.inject.Inject;
 import io.cdap.cdap.app.runtime.ProgramRuntimeService;
+import io.cdap.cdap.common.app.RunIds;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
 import io.cdap.cdap.proto.ProgramRunStatus;
 import io.cdap.cdap.proto.ProgramType;
 import io.cdap.cdap.proto.id.ProgramRunId;
-import org.eclipse.jetty.util.ConcurrentHashSet;
+import org.apache.twill.common.Threads;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.util.Comparator;
 import java.util.List;
-import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
 
 /**
- * Helper class to maintain and return total number of launching and running run-records.
+ * Maintain and return total number of launching and running run-records.
+ * This class is used by flow-control mechanism for launch requests.
+ * It also has a cleanup mechanism to automatically remove old (i.e., configurable) entries from the counter as a
+ * safe-guard mechanism.
  */
-public class RunRecordCounter {
-  private static final Logger LOG = LoggerFactory.getLogger(RunRecordCounter.class);
+public class RunRecordCounterService extends AbstractScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(RunRecordCounterService.class);
 
   /**
    * Contains ProgramRunIds of runs that have been accepted, but have not been added to metadata store plus
    * all run records with {@link ProgramRunStatus#PENDING} or {@link ProgramRunStatus#STARTING} status.
    */
-  private final Set<ProgramRunId> launchingSet;
+  private final PriorityBlockingQueue<ProgramRunId> launchingQueue;
   private final ProgramRuntimeService runtimeService;
+  private final long entryRetentionSec;
+  private final CConfiguration cConf;
+  private ScheduledExecutorService executor;
 
-  public RunRecordCounter(ProgramRuntimeService runtimeService) {
+  @Inject
+  public RunRecordCounterService(CConfiguration cConf, ProgramRuntimeService runtimeService) {
     this.runtimeService = runtimeService;
+    this.cConf = cConf;
+
+    launchingQueue = new PriorityBlockingQueue<>(100, new Comparator<ProgramRunId>() {
```

Comment:
nit: 128, use power of 2 whenever possible

Reply:
Please make a constant (not configurable, just class constant) for 100, this would explain what this number is. 
Also Consider using `Comparator.comparing(o -> RunIds.getTime(o.getRun(), TimeUnit.MILLISECONDS))`